base/loading.jl

```diff
@@ -511,26 +592,55 @@ module_uuid(m::Module) = ccall(:jl_module_uuid, UInt64, (Any,), m)
 
 isvalid_cache_header(f::IOStream) = 0 != ccall(:jl_read_verify_header, Cint, (Ptr{Void},), f.ios)
 
-function cache_dependencies(f::IO)
-    modules = Tuple{Symbol,UInt64}[]
-    files = Tuple{String,Float64}[]
+function parse_cache_header(f::IO)
+    modules = Dict{Symbol,UInt64}()
     while true
         n = ntoh(read(f, Int32))
         n == 0 && break
-        push!(modules,
-              (Symbol(read(f, n)), # module symbol
-               ntoh(read(f, UInt64)))) # module UUID (timestamp)
+        sym = Symbol(read(f, n)) # module symbol
+        uuid = ntoh(read(f, UInt64)) # module UUID (mostly just a timestamp)
+        modules[sym] = uuid
     end
-    read(f, Int64) # total bytes for file dependencies
+    totbytes = ntoh(read(f, Int64)) # total bytes for file dependencies
+    # read the list of files
+    files = Tuple{String,Float64}[]
     while true
         n = ntoh(read(f, Int32))
         n == 0 && break
+        totbytes -= 4 + n + 8
+        if n < 0 # probably means this wasn't a valid file to be read by Base.parse_cache_header
+            error("EOF while reading cache header")
+        end
         push!(files, (String(read(f, n)), ntoh(read(f, Float64))))
     end
+    @assert totbytes == 4 "header of cache file appears to be corrupt"
```

Comment:
Good point, the earlier errors in this function should also have more information if at all possible.

> Assertions are a separate universe from normal error messages.

I get that they're not intended to be user-friendly, but given

> any assertion failure is pretty much guaranteed to require you to dig through the source code and debug, because it is an error by the programmer and not an error by the user

I don't understand why you want to leave landmines with so little information for yourself if you're bothering to check something at all. If you have any doubt that you might be wrong and it might ever fail, why the resistance to leaving more breadcrumbs to help future you out and have an easier time of the digging and debugging? Especially if something only happens rarely or on a specific remote system reported by a user where you can't immediately reproduce.